src/main/scala/me/jeffmay/neo4j/client/ws/WSNeo4jClient.scala

```diff
@@ -0,0 +1,208 @@
+    // TODO: Allow configuration per-request
+    val jsonBody = Json.toJson(RawNeo4jCypherRequest(statements, includeStats = Some(config.includeStatsByDefault)))
+    val request = http("/db/data/transaction/commit")
+    whenInitialized {
+      request.postAndCheckStatus(Some(jsonBody)).flatMap { resp =>
+        // TODO: Handle version / json format errors with common recovery code
+        val respBody = resp.json.as[RawTxnResponse]
+        if (respBody.isSuccessful) {
+          Future.successful(respBody.asCommitTxnResponse.get)
+        }
+        else {
+          Future.failed(statusCodeException(request, Some(jsonBody), resp, respBody.errors))
+        }
+      }
+    }
+  }
+
+  override def commitTxn(ref: TxnRef, statements: Seq[CypherStatement]): Future[CommittedTxnResponse] = {
+    // TODO: Allow configuration per-request
+    val jsonBody = Json.toJson(RawNeo4jCypherRequest(statements, includeStats = Some(config.includeStatsByDefault)))
+    val request = http("/db/data/transaction")
+    whenInitialized {
+      request.postAndCheckStatus(Some(jsonBody)).flatMap { resp =>
+        // TODO: Handle version / json format errors with common recovery code
+        val respBody = resp.json.as[RawTxnResponse]
+        if (respBody.isSuccessful) {
+          Future.successful(respBody.asCommitTxnResponse.get)
+        }
+        else {
+          Future.failed(statusCodeException(request, Some(jsonBody), resp, respBody.errors))
+        }
+      }
+    }
+  }
+}
+
+object WSNeo4jClient {
+
+  // TODO: Replace this with an actor to manage throttling / circuit breaker / backoff retry on unauthorized or connection issues
+
+  @volatile private var lastSuccessfulRequestWithCreds: Option[Neo4jBasicAuth] = None
+
+  // this is a very naive implementation of recovery that should evolve into a more robust / non-blocking solution
+  private def changePasswordWithCredentials(client: Neo4jClient, creds: Neo4jBasicAuth): Unit = {
+    blocking {
+      synchronized {
+        // TODO: Is username needed for this?
+        Await.result(client.changePassword(creds.password), 1.second)
+        // If no exception is thrown, let's update the creds
+        lastSuccessfulRequestWithCreds = Some(creds)
+      }
+    }
+  }
+
+  private def withVerifiedCredentials[T](client: Neo4jClient, creds: Neo4jBasicAuth)(f: => Future[T])
+    (implicit ec: ExecutionContext): Future[T] = {
+    f.recoverWith {
+      case resp: UnexpectedStatusException if resp.responseStatus == Status.UNAUTHORIZED =>
+        WSNeo4jClient.synchronized {
+          val currentCreds = lastSuccessfulRequestWithCreds
+          lastSuccessfulRequestWithCreds = None
+          if (!(currentCreds contains creds)) {
+            Future(changePasswordWithCredentials(client, creds)).flatMap(_ => f)
+          }
+          else {
+            Future.failed(resp)
+          }
+        }
+    }
+  }
```

Comment:
I would definitely rather use Akka for this then roll with this kind of shitty synchronization.

Basically, this is all because the Neo4j database starts up with a default username / password of neo4j / neo4j and you have to change it to make any future requests.

BUT if you make an unauthorized request more than once, it rate limits you to one request per 5 seconds until you are authenticated.

Now the tricky part. The tests all run in parallel, so they all make their first request against a database with the wrong password, then they all try to recover by changing the password... You can see where this issue arises. They all get rate-limit blocked and all the tests fail.